Initial ESP32-C3-DevKitM-1 board support

agners opened the pull request:
# What does this implement/fix? 

Add board support for ESP32-C3-DevKitM-1. This board has been added to `platform-espressif32` as well. Currently the Arduino SDK is not officially supported yet for this board. Manually adding and overriding `framework-arduinoespressif32` allows to test the board with pre-releases of the upcoming Arduino SDK ESP32 v2.0.0 version.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1217

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32-C3
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
esphome:
  name: esp32-c3
  platform: ESP32
  board: esp32-c3-devkitm-1
  platformio_options:
    platform: https://github.com/agners/platform-espressif32#v3.3.0-arduino-v2.0.0-rc1
    platform_packages:
      - framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git#2.0.0-alpha1
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

agners committed:
Add ESP32-C3-DevKitM-1 support

The board/pin dictionaries are also used to validate board
configuration. Move definitions into esphome/boards.py to separate
them from pin logic.

Introduce ESP32_C3_BASE_PINS/ESP32_C3_BOARD_PINS for ESP32-C3 boards.
This implementation does not introduce a new platform as ESP32-C3 is
supported by the same platform as ESP32 devices.

agners committed:
Support ESP32-C3 ADC1

ADC1 supports up to 5 ADC channels. Use custom factors since
attenuation seems to be different on ESP32-C3. The factors have
been calculated from manual measurements on a ESP32-C3-DevKit.

agners commented:
While ESP32-C3 is quite different from ESP32, I chose to implement it "just a board" and not a new platform, Eg.

```
esphome:
  platform: ESP32
  board: esp32-c3-devkitm-1
```

as opposed to

```
esphome:
  platform: ESP32-C3
  board: esp32-c3-devkitm-1
```

It does lead to a bit custom code to find the chip from the board file. But this aligns with how the board is treated in platform.io: It is part of the same platform and frameworks (unlike ESP8266, which is separate). We could implement it as a separate platform in ESPHome still, and link it to the same platform.io platform/framework  as ESP32, but this might get confusing over time.

A possible alternative would be to introduce a new `mcu` node:

```
esphome:
  platform: ESP32
  mcu: esp32c3
  board: esp32-c3-devkitm-1
```

Thoughts?

agners committed:
Fix linter errors

agners committed:
Fix pytest

stintel committed:
ESP32-C3 uses different flash pins

Signed-off-by: Stijn Tintel <stijn@linux-ipv6.be>

stintel commented:
Thanks again for your work on this!

I'm trying to test this but the pin validation is wrong. ESP32-C3 uses different pins for flash than ESP32. See stintel/esphome@310241843da5256a30180760fa52fe361f3b8d5b - this solves that for me, based on the recommendation in the datasheet not to use the flash pins for other uses.

Another issue which I mentioned in esphome/feature-requests#1217:
```
src/esphome/components/debug/debug_component.cpp:8:10: fatal error: rom/rtc.h: No such file or directory
 #include <rom/rtc.h>
          ^~~~~~~~~~~
```
Fix at stintel/esphome@87ba5f629cf6ffd540fe1fca0ac484513a4c363c.

You can pull my for-agners branch and add it to this PR, it contains those 2 fixes. With those 2 on top of your PR, and the yaml from esphome/feature-requests#1217, I can use the ESP32-C3-DevKitM-1.

agners committed:
Fix ESP32-C3 pins error message

agners commented:
Thanks for the update, added your pin validation.

Wrt rtc:
> Fix at stintel/esphome@2e7a8b3.

Your change is correct to support ESP32 on ESP-IDF, but for ESP32-C3  the debug component seems to need more changes: It seems that there is also `esp32c3/rom/rtc.h` which would be the proper header file for ESP32-C3. It contains different defines. I'd suggest to leave the debug component changes for a follow up PR.

digiblur commented:
I'll update the video description soon I see! I linked this in the video for now.
https://youtu.be/YjKEpQDj8S0

agners committed:
Fix pylint issue

jesserockz commented on esphome/core/config.py:
```diff
@@ -50,18 +50,21 @@
+        boardlist += list(boards.ESP32_C3_BOARD_PINS.keys())
     else:
         raise NotImplementedError
+    # for board in sorted([str(value) for value in boards]):
+    #    print(board)
```
Remove debug code?

agners committed:
Remove debug code

agners committed:
Resolve filename/loc from Exception handler

It seems that currently ESP32-C3 does not print a backtrace on
exceptions. But resolving program counter/return address already helps
to pinpoint crashes.

agners commented:
I also added filename/loc resolving on Exception (in RISC-V the registers are named slightly different, MEPC and RA). Currently there seems to be no backtrace with RISC-V, not sure why that is. An Exception now looks like this:

```
[08:34:28]Guru Meditation Error: Core  0 panic'ed (Load access fault). Exception was unhandled.
[08:34:28]
[08:34:28]Core  0 register dump:
[08:34:28]MEPC    : 0x420228c8  RA      : 0x4202292a  SP      : 0x3fcad480  GP      : 0x3fc8da00  
INFO Need to fetch platformio IDE-data, please stand by
INFO Running:  platformio run -d config/esp32-c3 -t idedata
WARNING Decoded 0x420228c8: String::buffer() const at /root/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.h:322
 (inlined by) String::c_str() const at /root/.platformio/packages/framework-arduinoespressif32/cores/esp32/WString.h:230
 (inlined by) AsyncWebServerRequest::_removeNotInterestingHeaders() at /workspaces/esphome/config/esp32-c3/.piolibdeps/esp32-c3/ESPAsyncWebServer-esphome/src/WebRequest.cpp:184
[08:34:29]TP      : 0x3fc9598c  T0      : 0x7f7f7fff  T1      : 0x4038bf08  T2      : 0xffffffff  
[08:34:29]S0/FP   : 0x3fc9c7ec  S1      : 0x3fc95e34  A0      : 0x00000001  A1      : 0x815c31cf  
[08:34:29]A2      : 0x3fc9c7ec  A3      : 0x00000001  A4      : 0x3fc95000  A5      : 0x3fc95e34  
[08:34:29]A6      : 0x70626577  A7      : 0x2f656761  S2      : 0x3fcae22c  S3      : 0x3fcae23c  
[08:34:29]S4      : 0x3fc95e34  S5      : 0x00000000  S6      : 0x00000000  S7      : 0x00000000  
[08:34:29]S8      : 0x00000000  S9      : 0x00000000  S10     : 0x00000000  S11     : 0x00000000  
[08:34:29]T3      : 0x2e303d71  T4      : 0x3b6c6d78  T5      : 0x2f6e6f69  T6      : 0x74616369  
[08:34:29]MSTATUS : 0x00001881  MTVEC   : 0x40380001  MCAUSE  : 0x00000005  MTVAL   : 0x815c31de  
[08:34:29]MHARTID : 0x00000000  
[08:34:29]
[08:34:29]Stack memory:
[08:34:29]3fcad480: 0xc49f2c00 0x06ff0000 0x3fc95e34 0x00000000 0x00000000 0x00000000 0x00000000 0x03024f2a
[08:34:29]3fcad4a0: 0x00000000 0x00000000 0x00000000 0x04000000 0x00000051 0x00000000 0x3fcae22c 0x3fcae260
[08:34:29]3fcad4c0: 0x3fcae69f 0x00000002 0x3fcae22c 0x420229d2 0x00000000 0x00000000 0x420204fa 0x3fcae260
[08:34:29]3fcad4e0: 0x3fcae69f 0x3fcae260 0x3fcae69f 0x3fcae260 0x3fcae69f 0x00000002 0x00000001 0x42022ab8
[08:34:29]3fcad500: 0x40388e0c 0x40388e36 0x42020524 0x420206d8 0x3fc9598c 0x4005890e 0x0000000f 0x0000000c
[08:34:29]3fcad520: 0x00000000 0x0000000c 0x3fc95e20 0x4038b9ea 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad540: 0x00000000 0x3fca928c 0x3fcae158 0x42022d3c 0x00000000 0x0000000c 0x0000000c 0x4201a62c
[08:34:29]3fcad560: 0x00000143 0x3fcae55e 0x3fcae158 0x3fcae22c 0x00000000 0x0000000c 0x00000001 0x4201a688
[08:34:29]3fcad580: 0x00000000 0x00000000 0x3fcadd58 0x4201a6cc 0x00000000 0x00000000 0x00000000 0x4201a856
[08:34:29]3fcad5a0: 0x00000000 0x00000000 0x00000000 0x4201a8c8 0x00000000 0x00000000 0x00000000 0x3fcadd58
[08:34:29]3fcad5c0: 0x00000000 0x00000000 0x00000000 0x40388e82 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad5e0: 0x00000000 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0x27146414 0x01010107 0x00000001 0x00000000
[08:34:29]3fcad600: 0x35d1cfdc 0x00000000 0x3f15c000 0x3f800000 0x3fc00000 0x4b000000 0xcb000000 0xff800000
[08:34:29]3fcad620: 0x7f800000 0xbf800000 0x7149f2ca 0x3fb8aa00 0x36eca570 0x3e800000 0x3eaaaaab 0x3f000000
[08:34:29]3fcad640: 0x3fb8aa3b 0x4b800000 0x3e53f142 0x3e6c3255 0x3e8ba305 0x3edb6db7 0x3f19999a 0x3f763800
[08:34:29]3fcad660: 0x3f76384f 0x369dc3a0 0x3338aa3c 0x0da24260 0x3f317200 0x3f317218 0x35bfbe8c 0x3331bb4c
[08:34:29]3fcad680: 0x35ddea0e 0x388ab355 0x3b360b61 0x3e2aaaab 0x4c000000 0xf149f2ca 0x8da24260 0x33000000
[08:34:29]3fcad6a0: 0x00000000 0x00000000 0x3fc00000 0x00000000 0x40300000 0x00000000 0x3fe00000 0x00000000
[08:34:29]3fcad6c0: 0x3ff80000 0x636f4361 0x3fd287a7 0x8b60c8b3 0x3fc68a28 0x509f79fb 0x3fd34413 0x00000000
[08:34:29]3fcad6e0: 0x401c0000 0x00000000 0x40140000 0x00000000 0x43500000 0x00000018 0x00000000 0x4c507a03
[08:34:29]3fcad700: 0xa5a5a5a5 0xa5a5a5a5 0xa5a5a5a5 0x00000154 0x3fcad500 0x3fc95e34 0x3fc91c54 0x3fc91c54
[08:34:29]3fcad720: 0x3fcad710 0x3fc91c4c 0x00000016 0x3fca9660 0x3fca9660 0x3fcad710 0x00000000 0x00000003
[08:34:29]3fcad740: 0x3fca970c 0x6e797361 0x63745f63 0x3fca0070 0x00ca9680 0x00000000 0x3fcad700 0x00000003
[08:34:29]3fcad760: 0x00000000 0x3fc9cea4 0x420330be 0x00000000 0x3fc9664c 0x3fc966b4 0x3fc9671c 0x00000000
[08:34:29]3fcad780: 0x00000000 0x00000001 0x00000000 0x00000000 0x00000000 0x420a8ba8 0x00000000 0x00000000
[08:34:29]3fcad7a0: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad7c0: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad7e0: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad800: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad820: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad840: 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000 0x00000000
[08:34:29]3fcad860: 0x3fca0000 0x0000000c 0x3fc9c98c 0x3fcad888 0x3fcad860 0x0000000c 0x3fc9c9c4 0x3fcad868
[08:34:29]
[08:34:29]
[08:34:29]
[08:34:29]ELF file SHA256: 0000000000000000
[08:34:29]
[08:34:29]Rebooting...
[08:34:29]ESP-ROM:esp32c3-api1-20210207
...
```

OttoWinter commented:
Hmm yeah that looks similar to an ESP8266 dump (just spits out whole stack) - we can use a similar approach for ESP32-C3 to just decode all addresses that _look_ like code addresses (on ESP8266 it's easy because code is always in certain fixed address ranges).

But the PC decoding alone is already quite good. We can always improve things later.

markusg1234 commented:
Looking good, have now have ADC working. 
I tried to use deep sleep and had the following errors using,
deep_sleep:
 run_duration: 60sec
 sleep_duration: 60min

How do I declare ext0 and ext1?

Regards,
Markus

INFO Reading configuration /config/esphome/esp32-c.yaml...
INFO Generating C++ source...
INFO Compiling app...
INFO Running:  platformio run -d /config/esphome/esp32-c
Processing esp32-c (board: esp32-c3-devkitm-1; framework: arduino; platform: https://github.com/agners/platform-espressif32#v3.3.0-arduino-v2.0.0-rc1)
--------------------------------------------------------------------------------
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
PACKAGES: 
 - framework-arduinoespressif32 0.0.0+sha.5d9b98c 
 - tool-esptoolpy 1.30100.210531 (3.1.0) 
 - toolchain-riscv-esp 1.80400.0 (8.4.0)
Library Manager: Installing Hash
Library Manager: Already installed, built-in library
Dependency Graph
|-- <AsyncTCP-esphome> 1.2.2
|-- <ESPmDNS> 1.0
|   |-- <WiFi> 1.0
|-- <Update> 1.0
|-- <WiFi> 1.0
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/deep_sleep/deep_sleep_component.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/switch/switch.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/wifi/wifi_component.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/wifi/wifi_component_esp32.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/components/wifi_signal/wifi_signal_sensor.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/core/application.cpp.o
Compiling /data/esp32-c/.pioenvs/esp32-c/src/esphome/core/component.cpp.o
src/esphome/components/deep_sleep/deep_sleep_component.cpp: In member function 'void esphome::deep_sleep::DeepSleepComponent::begin_sleep(bool)':
src/esphome/components/deep_sleep/deep_sleep_component.cpp:78:5: error: 'esp_sleep_enable_ext0_wakeup' was not declared in this scope
     esp_sleep_enable_ext0_wakeup(gpio_num_t((*this->wakeup_pin_)->get_pin()), level);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/deep_sleep/deep_sleep_component.cpp:78:5: note: suggested alternative: 'esp_sleep_enable_wifi_wakeup'
     esp_sleep_enable_ext0_wakeup(gpio_num_t((*this->wakeup_pin_)->get_pin()), level);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
     esp_sleep_enable_wifi_wakeup
src/esphome/components/deep_sleep/deep_sleep_component.cpp:81:5: error: 'esp_sleep_enable_ext1_wakeup' was not declared in this scope
     esp_sleep_enable_ext1_wakeup(this->ext1_wakeup_->mask, this->ext1_wakeup_->wakeup_mode);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/deep_sleep/deep_sleep_component.cpp:81:5: note: suggested alternative: 'esp_sleep_enable_wifi_wakeup'
     esp_sleep_enable_ext1_wakeup(this->ext1_wakeup_->mask, this->ext1_wakeup_->wakeup_mode);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
     esp_sleep_enable_wifi_wakeup
*** [/data/esp32-c/.pioenvs/esp32-c/src/esphome/components/deep_sleep/deep_sleep_component.cpp.o] Error 1
========================== [FAILED] Took 2.48 seconds ==========================
